sys/net/gnrc/link_layer/lwmac/lwmac.c

```diff
@@ -0,0 +1,733 @@
+/*
+ * Copyright (C) 2015 Daniel Krebs
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @{
+ * @ingroup     net_lwmac
+ * @file
+ * @brief       Implementation of the LWMAC protocol
+ *
+ * @author      Daniel Krebs <github@daniel-krebs.net>
+ * @}
+ */
+
+#include <errno.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "kernel.h"
+#include "msg.h"
+#include "thread.h"
+#include "timex.h"
+#include "priority_queue.h"
+#include "periph/rtt.h"
+#include "net/gnrc.h"
+#include "net/gnrc/lwmac/lwmac.h"
+#include "net/gnrc/lwmac/packet_queue.h"
+#include "net/gnrc/lwmac/timeout.h"
+#include "net/gnrc/lwmac/tx_state_machine.h"
+#include "net/gnrc/lwmac/rx_state_machine.h"
+
+#define ENABLE_DEBUG    (1)
+#include "debug.h"
+
+#define LOG_LEVEL LOG_INFO
+#include "log.h"
+
+#undef LOG_ERROR
+#undef LOG_WARNING
+#undef LOG_INFO
+#undef LOG_DEBUG
+
+#define LOG_ERROR(...) LOG(LOG_ERROR, "ERROR: [lwmac] " __VA_ARGS__)
+#define LOG_WARNING(...) LOG(LOG_WARNING, "WARNING: [lwmac] " __VA_ARGS__)
+#define LOG_INFO(...) LOG(LOG_INFO, "[lwmac] " __VA_ARGS__)
+#define LOG_DEBUG(...) LOG(LOG_DEBUG, "[lwmac] " __VA_ARGS__)
+
+/******************************************************************************/
+
+/* Internal state of lwMAC */
+static lwmac_t lwmac = LWMAC_INIT;
+
+/******************************************************************************/
+
+static bool lwmac_update(void);
+static void lwmac_set_state(lwmac_state_t newstate);
+static void lwmac_schedule_update(void);
+static bool lwmac_needs_update(void);
+static void rtt_handler(uint32_t event);
+
+/******************************************************************************/
+
+int _get_dest_address(gnrc_pktsnip_t* pkt, uint8_t* pointer_to_addr[])
+{
+    int res;
+    gnrc_netif_hdr_t* hdr;
+
+    if(!pkt)
+        return -ENODEV;
+
+    hdr = (gnrc_netif_hdr_t*) pkt->data;
+    if( (res = hdr->dst_l2addr_len) <= 0)
+        return -ENOENT;
+
+    *pointer_to_addr = gnrc_netif_hdr_get_dst_addr(hdr);
+    return res;
+}
+
+/******************************************************************************/
+
+/* Find a payload based on it's protocol type */
+void* _gnrc_pktbuf_find(gnrc_pktsnip_t* pkt, gnrc_nettype_t type)
+{
+    while(pkt != NULL)
+    {
+        if(pkt->type == type) {
+            return pkt->data;
+        }
+        pkt = pkt->next;
+    }
+    return NULL;
+}
+
+/******************************************************************************/
+
+bool _accept_packet(gnrc_pktsnip_t* pkt, lwmac_frame_type_t expected_type, lwmac_t* lwmac)
+{
+    gnrc_netif_hdr_t* netif_hdr;
+    lwmac_hdr_t* lwmac_hdr;
+    uint8_t* own_addr = (uint8_t*) &(lwmac->addr);
+    uint8_t* dst_addr;
+
+    netif_hdr = _gnrc_pktbuf_find(pkt, GNRC_NETTYPE_NETIF);
+    if(netif_hdr == NULL) {
+        LOG_WARNING("Reject packet: no NETIF header\n");
```

Comment:
Off course. I don't like that I made my own logging/debugging macros either. All debug messages will need a review when this is ready to be merged. For the time being these will stay as-is.